cpp/src/arrow/ipc/writer.h

```diff
@@ -104,8 +104,7 @@ class ARROW_EXPORT RecordBatchWriter {
   virtual Status WriteRecordBatch(
       const RecordBatch& batch,
       const std::shared_ptr<const KeyValueMetadata>& custom_metadata) {
-    return Status::NotImplemented(
-        "Write record batch with custom metadata not implemented");
+    return WriteRecordBatch(batch);
```

Comment:
In https://github.com/apache/arrow/pull/12812, @pitrou suggested to ignore the custom_metadata by default. By I somehow missed that part and the default implementation is throwing an error. When I was adding this API in pyarrow, I found one test case failed (https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_csv.py#L1895). And in this csv test case, it uses the `CsvWriter` which doesn't implement this API signature.

Since storing custom metadata is an IPC feature, probably other serialization won't support this so far, I realize probably we should switch the default implementation to discarding the provided custom_metadata.

Reply:
I think it's ok to return `Status::NotImplemented`, but of course as long as there are no regressions in other components :-)

Reply:
It won't cause crash but this default implementation (ignoring custom_metadata if not supported) seems making higher level binding easier to write. Otherwise, for example, pyarrow's implementation will need to choose which API to use like this:
```
def write_batch(self, batch, custom_metadata=None):
   if (custom_metadata):
     WriteRecordBatch(batch, custom_metadata);
   else:
     WriteRecordBatch(batch);
```
It seems this will be a pattern for all higher level bindings, which I think probably indicates this could be handled in lower level API, so that higher level binding could always be implemented like this:
```
 def write_batch(self, batch, custom_metadata=None):
     WriteRecordBatch(batch, custom_metadata);
```
What do you think?

Reply:
Well, the default implementation can then be:
```c++
  virtual Status WriteRecordBatch(
      const RecordBatch& batch,
      const std::shared_ptr<const KeyValueMetadata>& custom_metadata) {
    if (custom_metadata == nullptr) {
      return WriteRecordBatch(batch);
    }
    return Status::NotImplemented(
        "Write record batch with custom metadata not implemented");
  }
```

Reply:
Change the default implementation according to suggestion.